File: surf-api-core/surf-api-core-api/src/main/java/dev/slne/surf/surfapi/core/api/util/ById.java
package dev.slne.surf.surfapi.core.api.util;

import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
import org.jetbrains.annotations.NotNull;

@FunctionalInterface
public interface ById {

  static <T extends Enum<T> & ById> @NotNull Int2ObjectMap<T> build(@NotNull Class<T> clazz) {
    return Util.byIdMap(ById::id, clazz.getEnumConstants());
  }

  int id();

  @FunctionalInterface
  interface ByByteId {

    byte id();
  }
}

File: surf-api-core/surf-api-core-api/src/main/kotlin/dev/slne/surf/surfapi/core/api/util/ById.kt
package dev.slne.surf.surfapi.core.api.util

import it.unimi.dsi.fastutil.ints.Int2ObjectMap
import java.util.function.ToIntFunction

fun interface ById {
    fun id(): Int

    fun interface ByByteId {
        fun id(): Byte
    }

    companion object {
        fun <T> build(clazz: Class<T>): Int2ObjectMap<T> where T : Enum<T>, T : ById {
            return Util.byIdMap<T>(
                ToIntFunction { it.id() },
                clazz.getEnumConstants()
            )
        }
    }
}
